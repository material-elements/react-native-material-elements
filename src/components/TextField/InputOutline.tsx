import React, { useMemo } from 'react';
import { StyleSheet, View } from 'react-native';
import { useThemeColorsSelector } from '../../libraries';
import { OutlineProps } from './Input.types';
import { inputOutlineVariationStyles, outlineStyles } from './TextField.style';

export const Outline = React.forwardRef<View, OutlineProps>(
  (
    {
      error,
      style,
      isFocused,
      activeColor,
      errorColor,
      editable,
      ignoreOpacityOnNonEditable,
      square,
      variant = 'outlined',
      borderRadius,
      ...props
    },
    ref,
  ) => {
    const themeColors = useThemeColorsSelector();

    const outlineGeneratedStyles = useMemo(() => {
      return {
        ...inputOutlineVariationStyles(variant, themeColors),
        ...outlineStyles({
          error,
          errorColor,
          isFocused,
          activeColor,
          colors: themeColors,
          editable,
          variant,
          ignoreOpacityOnNonEditable,
          square,
          borderRadius,
        }),
      };
    }, [
      error,
      errorColor,
      isFocused,
      activeColor,
      themeColors,
      editable,
      variant,
      ignoreOpacityOnNonEditable,
      square,
      borderRadius,
    ]);

    return <View ref={ref} style={StyleSheet.flatten([outlineGeneratedStyles, style])} {...props} />;
  },
);
Outline.displayName = 'TextFiledOutline';
