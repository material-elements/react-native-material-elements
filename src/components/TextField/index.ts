export * from './TextField';
export * from './BaseInput';
export * from './InputLabel';
export * from './IconInput';
export * from './OtpInput';
