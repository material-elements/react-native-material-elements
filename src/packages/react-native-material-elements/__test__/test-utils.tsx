import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';
import { ThemeProviderProps } from '../src/libraries/types';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  themeProps?: Partial<ThemeProviderProps>;
}

export const ThemeWrapper: React.FC<ThemeProviderProps> = ({ children, ...props }) => {
  return <ThemeProvider {...props}>{children}</ThemeProvider>;
};

export const customRender = (ui: React.ReactElement, options?: CustomRenderOptions) => {
  const { themeProps, ...rest } = options || {};
  return render(ui, {
    wrapper: wrapperProps => <ThemeWrapper {...themeProps}>{wrapperProps.children}</ThemeWrapper>,
    ...rest,
  });
};

export * from '@testing-library/react-native';
export { customRender as render };
