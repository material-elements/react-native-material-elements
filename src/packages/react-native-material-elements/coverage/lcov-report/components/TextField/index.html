
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/TextField</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> components/TextField</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">85.85% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>85/99</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">70.9% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>39/55</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">81.81% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>18/22</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">85.85% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>85/99</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="BaseInput.tsx"><a href="BaseInput.tsx.html">BaseInput.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="IconInput.tsx"><a href="IconInput.tsx.html">IconInput.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file high" data-value="InputLabel.tsx"><a href="InputLabel.tsx.html">InputLabel.tsx</a></td>
	<td data-value="84.61" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 84%"></div><div class="cover-empty" style="width: 16%"></div></div>
	</td>
	<td data-value="84.61" class="pct high">84.61%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	<td data-value="84.61" class="pct high">84.61%</td>
	<td data-value="13" class="abs high">11/13</td>
	</tr>

<tr>
	<td class="file high" data-value="InputOutline.tsx"><a href="InputOutline.tsx.html">InputOutline.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	</tr>

<tr>
	<td class="file medium" data-value="OtpInput.tsx"><a href="OtpInput.tsx.html">OtpInput.tsx</a></td>
	<td data-value="78.72" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 78%"></div><div class="cover-empty" style="width: 22%"></div></div>
	</td>
	<td data-value="78.72" class="pct medium">78.72%</td>
	<td data-value="47" class="abs medium">37/47</td>
	<td data-value="68" class="pct medium">68%</td>
	<td data-value="25" class="abs medium">17/25</td>
	<td data-value="76.92" class="pct medium">76.92%</td>
	<td data-value="13" class="abs medium">10/13</td>
	<td data-value="78.72" class="pct medium">78.72%</td>
	<td data-value="47" class="abs medium">37/47</td>
	</tr>

<tr>
	<td class="file high" data-value="TextFieldEndAdornment.tsx"><a href="TextFieldEndAdornment.tsx.html">TextFieldEndAdornment.tsx</a></td>
	<td data-value="83.33" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 83%"></div><div class="cover-empty" style="width: 17%"></div></div>
	</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	<td data-value="70.58" class="pct medium">70.58%</td>
	<td data-value="17" class="abs medium">12/17</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	</tr>

<tr>
	<td class="file high" data-value="constants.ts"><a href="constants.ts.html">constants.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-09-20T12:52:29.060Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    