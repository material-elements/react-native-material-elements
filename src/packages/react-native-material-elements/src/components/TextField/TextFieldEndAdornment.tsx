import React from 'react';
import { ActivityIndicatorProps } from 'react-native';
import { ActivityIndicator } from '../ActivityIndicator';
import { Box } from '../Box';
import { BoxProps } from '../types';

export interface TextFieldEndAdornmentProps {
  loading?: boolean;
  endAdornment?: React.ReactNode;
  showLoadingIndicatorWhenFocused?: boolean;
  loadingIndicatorProps?: ActivityIndicatorProps;
  isFocused?: boolean;
  endAdornmentContainerProps?: Omit<BoxProps, 'children'>;
  testID?: string;
}

const TextFieldEndAdornment: React.FC<TextFieldEndAdornmentProps> = ({
  loading,
  endAdornment,
  showLoadingIndicatorWhenFocused,
  isFocused,
  loadingIndicatorProps,
  endAdornmentContainerProps,
  testID,
}) => {
  if (!loading && !endAdornment) {
    return null;
  }

  let element: React.ReactNode = null;

  if (endAdornment) {
    element = endAdornment;
  } else if ((showLoadingIndicatorWhenFocused && isFocused && loading) || (!showLoadingIndicatorWhenFocused && loading)) {
    element = <ActivityIndicator {...loadingIndicatorProps} />;
  }

  if (!element) {
    return null;
  }

  return (
    <Box sx={{ me: 8, ms: 8 }} {...endAdornmentContainerProps} testID={testID}>
      {element}
    </Box>
  );
};

export default TextFieldEndAdornment;
