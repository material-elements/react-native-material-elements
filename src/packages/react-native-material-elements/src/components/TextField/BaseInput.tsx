import React from 'react';
import { StyleSheet, TextInput } from 'react-native';
import { useThemeColorsSelector } from '../../libraries';
import { BaseInputProps } from './Input.types';
import { baseInputStyles } from './TextField.style';

export const BaseInput = React.forwardRef<TextInput, BaseInputProps>(({ style, variant, height, ...props }, ref) => {
  const themeColors = useThemeColorsSelector();

  return (
    <TextInput
      ref={ref}
      style={StyleSheet.flatten([baseInputStyles({ colors: themeColors, variant, height }), style])}
      {...props}
    />
  );
});
