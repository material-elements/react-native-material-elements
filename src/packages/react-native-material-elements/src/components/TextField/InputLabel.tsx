import React, { useMemo, useState } from 'react';
import { Animated, LayoutChangeEvent, LayoutRectangle } from 'react-native';
import { useThemeColorsSelector } from '../../libraries';
import { Text } from '../Typography';
import { InputLabelProps } from './Input.types';
import { labelTextStyles, labelTransformStyle } from './TextField.style';
import { PLACEHOLDER_OUTLINE_LEFT_POSITION, TEXT_FONT_DEFAULT_HEIGHT } from './constants';

export const InputLabel: React.FC<InputLabelProps> = function ({
  placeholder,
  labelAnimatedValue,
  editable,
  translateYAnimatedPosition,
  placeholderLeftPosition,
  labelContainerStyles,
  style,
  error,
  errorColor,
  ignoreOpacityOnNonEditable,
  variant = 'outlined',
  ...props
}) {
  const themeColors = useThemeColorsSelector();

  const [textLayoutRect, setTextLayoutRect] = useState<LayoutRectangle>();
  const textHeight = textLayoutRect?.height ?? TEXT_FONT_DEFAULT_HEIGHT;

  const styles = useMemo(
    () =>
      labelTransformStyle({
        colors: themeColors,
        textHeight,
        translateYAnimatedPosition,
        labelAnimatedValue,
        variant,
        placeholderLeftPosition: placeholderLeftPosition ?? PLACEHOLDER_OUTLINE_LEFT_POSITION,
      }),
    [themeColors, textHeight, translateYAnimatedPosition, labelAnimatedValue, variant, placeholderLeftPosition],
  );

  const labelStyles = useMemo(
    () => labelTextStyles({ colors: themeColors, variant, ignoreOpacityOnNonEditable, error, errorColor }),
    [themeColors, variant, ignoreOpacityOnNonEditable, error, errorColor],
  );

  const onTextLayoutHandler = (event: LayoutChangeEvent) => {
    const { layout } = event.nativeEvent;
    setTextLayoutRect(layout);
  };

  return (
    <Animated.View style={[styles, labelContainerStyles]}>
      <Text variation="h4" onLayout={onTextLayoutHandler} disabled={editable} style={[labelStyles, style]} {...props}>
        {placeholder}
      </Text>
    </Animated.View>
  );
};

InputLabel.displayName = 'InputLabel';
