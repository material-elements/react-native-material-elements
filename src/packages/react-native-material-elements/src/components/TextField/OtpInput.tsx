import React, { useEffect, useRef, useState } from 'react';
import {
  ColorValue,
  NativeSyntheticEvent,
  StyleSheet,
  TextInput,
  TextInputKeyPressEventData,
  TextInputProps,
  TextStyle,
  View,
  ViewStyle,
} from 'react-native';
import { useThemeColorsSelector } from '../../libraries';
import { VariantTypes } from '../../utils';
import { getOtpInputStyles } from './TextField.style';

export type OtpInputVariation = 'outlined' | 'underlined';
export type OTPInputProps = Omit<TextInputProps, 'onChange' | 'value' | 'onChangeText' | 'defaultValue'> & {
  /**
   * Total number of OTP input boxes.
   * Determines the length of the OTP code.
   */
  length: number;
  /**
   * Callback that receives the complete OTP string as user inputs values.
   */
  onChange?: (otp: string) => void;
  /**
   * Visual style variant (e.g., 'filled', 'outlined') to customize the appearance.
   */
  variant?: VariantTypes;
  /**
   * Styles applied to each individual OTP input box.
   */
  inputStyles?: TextStyle;
  /**
   * Styles applied to the container wrapping all OTP input boxes.
   */
  otpContainerStyles?: ViewStyle;
  /**
   * If true, renders each OTP input as a square.
   * Useful for consistent design across platforms.
   */
  square?: boolean;
  /**
   * If true, applies error styles to the inputs (e.g., red border).
   * Typically used for validation feedback.
   */
  error?: boolean;
  /**
   * Specifies the variation of the OTP input component.
   * Can be used to switch between different UI behavior or styles.
   */
  variation?: OtpInputVariation;
  /**
   * The active tint color for focused or filled input boxes.
   */
  tintColor?: ColorValue;
  /**
   * The inactive tint color for empty or unfocused input boxes.
   */
  offTintColor?: ColorValue;
  /**
   * Prefilled value for the OTP input.
   * Can be a string (e.g., "123") or number (e.g., 123) and will be split across input boxes.
   */
  defaultValue?: number | string;
};

export const OTPInput: React.FC<OTPInputProps> = ({
  length,
  onChange,
  inputStyles,
  otpContainerStyles,
  tintColor,
  offTintColor,
  defaultValue,
  square = false,
  error = false,
  variant = 'secondary',
  variation = 'outlined',
  testID = 'otp-input',
  ...props
}) => {
  const [otp, setOtp] = useState<string[]>(Array(length).fill(''));
  const [focusedIndex, setFocusedIndex] = useState<number | null>(null);
  const inputs = useRef<TextInput[]>([]);

  const isUnderLineInput = variation === 'underlined';

  const themeColors = useThemeColorsSelector();

  const handleChangeText = (text: string, index: number) => {
    if (text.length > 1) {
      return;
    }

    const newOtp = [...otp];
    newOtp[index] = text;
    setOtp(newOtp);
    if (onChange) {
      onChange(newOtp.join(''));
    }

    if (text && index < length - 1) {
      inputs.current[index + 1]?.focus();
    }
  };

  const handleKeyPress = (e: NativeSyntheticEvent<TextInputKeyPressEventData>, index: number) => {
    if (e.nativeEvent.key === 'Backspace') {
      const newOtp = [...otp];
      newOtp[index] = '';
      setOtp(newOtp);
      if (onChange) {
        onChange(newOtp.join(''));
      }

      if (index > 0) {
        inputs.current[index - 1]?.focus();
      }
    }
  };

  const handleFocus = (index: number) => {
    setFocusedIndex(index);
  };

  const handleBlur = () => {
    setFocusedIndex(null);
  };

  useEffect(() => {
    setOtp(Array(length).fill(''));
  }, [length]);

  useEffect(() => {
    if (defaultValue) {
      const defaultValueLength = defaultValue.toString();

      if (defaultValueLength.length > length) {
        throw Error('Default value must be equal or less then otp length');
      }

      const otpArray = defaultValueLength.split('').concat(Array(length - defaultValueLength.length).fill(''));
      setOtp(otpArray);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [defaultValue]);

  const renderInputs = () => {
    return otp.map((_, index) => (
      <TextInput
        key={index}
        style={StyleSheet.flatten([
          styles.input,
          getOtpInputStyles({
            length,
            colors: themeColors,
            variant,
            isFocused: focusedIndex === index,
            square,
            error,
            tintColor,
            offTintColor,
          }),
          inputStyles,
          isUnderLineInput && styles.underLineInput,
        ])}
        keyboardType="number-pad"
        maxLength={1}
        ref={ref => (inputs.current[index] = ref!)}
        onChangeText={text => handleChangeText(text, index)}
        onKeyPress={e => handleKeyPress(e, index)}
        value={otp[index]}
        onFocus={() => handleFocus(index)}
        onBlur={handleBlur}
        selectTextOnFocus
        testID={`${testID}-${index}`}
        {...props}
      />
    ));
  };

  return <View style={StyleSheet.flatten([styles.container, otpContainerStyles])}>{renderInputs()}</View>;
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  input: {
    height: 50,
    borderWidth: 0.8,
    color: 'red',
    textAlign: 'center',
    marginHorizontal: 5,
    fontSize: 18,
  },
  underLineInput: {
    borderRadius: 0,
    borderTopWidth: 0,
    borderRightWidth: 0,
    borderLeftWidth: 0,
  },
});
