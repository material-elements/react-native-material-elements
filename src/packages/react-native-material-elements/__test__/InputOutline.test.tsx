import { Outline } from '../src/components/TextField/InputOutline';
import { render } from './test-utils';

describe('InputOutline component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with default props', () => {
    const { toJSON } = render(<Outline />);
    expect(toJSON()).toMatchSnapshot();
  });
});
