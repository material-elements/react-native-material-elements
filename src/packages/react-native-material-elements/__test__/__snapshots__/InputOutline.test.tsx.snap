// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputOutline component should render correctly with default props 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "#BDBDBD",
      "borderRadius": undefined,
      "borderWidth": 0.6,
      "display": "flex",
      "flexDirection": "row",
      "opacity": 0.6,
      "paddingHorizontal": 14,
      "position": "relative",
      "width": "100%",
    }
  }
/>
`;
